### PR TITLE
Add analyzer/fixer to update expressions like `Array.Empty<T>()` over to using the `[]` collection expression.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Draft while waiting for https://github.com/dotnet/roslyn/pull/69219 to go in.

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -12,33 +12,30 @@
 using Microsoft.CodeAnalysis.LanguageService;
 using Microsoft.CodeAnalysis.UseCollectionInitializer;
 
-namespace Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerDiagnosticAnalyzer.cs:15]: review with whitespasce off.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForArrayCodeFixProvider.cs [hunk 2]
```diff
@@ -53,7 +54,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
         foreach (var diagnostic in diagnostics.OrderByDescending(d => d.Location.SourceSpan.Start))
         {
-            var expression = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+            var expression = diagnostic.AdditionalLocations[0].FindNode(getInnermostNodeForTie: true, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForArrayCodeFixProvider.cs:57]: this was a bug before, but was not something you could generally notice as the location was the first token of these nodes and asking for the node would give you teh same thing.  but htis is more correct/safer, so i updated.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForArrayCodeFixProvider.cs [hunk 3]
```diff
@@ -79,40 +80,103 @@ void RewriteInitializerExpression(InitializerExpressionSyntax initializer)
                 initializer,
                 (current, _) => ConvertInitializerToCollectionExpression(
                     (InitializerExpressionSyntax)current,
-                    IsOnSingleLine(initializer)));
+                    IsOnSingleLine(sourceText, initializer)));
+        }
+
+        bool ShouldReplaceExistingExpressionEntirely(ExpressionSyntax explicitOrImplicitArray, InitializerExpressionSyntax initializer)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForArrayCodeFixProvider.cs:86]: this is from https://github.com/dotnet/roslyn/pull/69278

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs [hunk 4]
```diff
@@ -243,7 +243,8 @@ class C
             FixedCode = """
                 class C
                 {
-                    object[] i = [
+                    object[] i =
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs:246]: this is from https://github.com/dotnet/roslyn/pull/69278

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForEmptyTests.cs [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+using System;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.UseCollectionExpression;
+using Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.UnitTests.UseCollectionExpression;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForEmptyTests.cs:14]: these are the new tests.

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal :)

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne ptal

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 6]
```diff
@@ -88,14 +88,14 @@ private static void AnalyzeArrayInitializer(SyntaxNodeAnalysisContext context)
             : initializer;
 
         if (!UseCollectionExpressionHelpers.CanReplaceWithCollectionExpression(
-                semanticModel, arrayCreationExpression, cancellationToken))
+                semanticModel, arrayCreationExpression, skipVerificationForReplacedNode: false, cancellationToken))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:91]: this is needed as the speculative system currently isn't ok with a node's kind changing this dramatically.  specifically from a method invocation to another expression form.  I wasn't comfortable mucking with the deep speculatrion logic, so easier to just not compare the bottommost nodes directly.

Comment (Cosifne) [reply src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:91]: 😄I don't have any context about this, but choose to trust you here

**Review by Cosifne — APPROVED** [review]
LGTM for the new analyzer/test part.

**Pull request merged**